### PR TITLE
Prevent Nav from Trapping while tabbing 

### PR DESCRIPTION
**Pull request by hqhhuang**
Bug/issue #, if applicable: rdar://94044029 ([AX] Localnav should not FocusTrap for documentation pages)

## Summary
For accessibility considerations, when we tab to the end of the nav menu, we should tab to the browser menu bar next (instead of tabbing to the beginning of the nav menu again).

## Dependencies
NA

## Testing
Steps:
1. Medium or small viewport, tab open the nav menu bar
2. tab through the items in the menu, at the end, focus should be brought to the browser address bar. It should not take you to the first link within our nav menu bar. 

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran `npm test`, and it succeeded
- [ ] Updated documentation if necessary - NA

### CODE REVIEW

**Comment by hqhhuang** [comment]
@swift-ci test

File: src/components/NavBase.vue [hunk 1]
```diff
@@ -315,9 +315,10 @@ export default {
         scrollLock.unlockScroll(this.$refs.tray);
       }
     },
-    onExpand() {
+    async onExpand() {
       this.$emit('open');
       // lock focus
+      await this.$nextTick();
```
Comment (dobromir-hristov) [inline src/components/NavBase.vue:321]: I know why we added this, but do we need it for this fix?

File: tests/unit/components/Asset.spec.js [hunk 2]
```diff
@@ -185,7 +185,7 @@ describe('Asset', () => {
       expect(asset.exists()).toBe(true);
       expect(asset.props()).toEqual({
         alt: image.alt,
-        loading: "lazy",
```
Comment (dobromir-hristov) [inline tests/unit/components/Asset.spec.js:188]: Lets revert these as they are not related to the PR.

File: src/utils/changeElementVOVisibility.js [hunk 3]
```diff
@@ -1,12 +1,44 @@
...
+import TabManager from 'docc-render/utils/TabManager';
+
+const PREFIX = 'data-original-';
+const ARIA = 'aria-hidden';
+const TABINDEX = 'tabindex';
+
+function setOriginalValue(element, prop) {
+  // make sure that prop isn't cached already
+  let originalValue = element.getAttribute(PREFIX + prop);
+  if (originalValue === undefined || originalValue === null || originalValue === '') {
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:19]: You can probably just check for `hasAttribute` instead?

File: src/utils/changeElementVOVisibility.js [hunk 4]
```diff
@@ -1,12 +1,44 @@
...
- * Copyright (c) 2021 Apple Inc. and the Swift project authors
+ * Copyright (c) 2022 Apple Inc. and the Swift project authors
  * Licensed under Apache License v2.0 with Runtime Library Exception
  *
  * See https://swift.org/LICENSE.txt for license information
  * See https://swift.org/CONTRIBUTORS.txt for Swift project authors
 */
+import TabManager from 'docc-render/utils/TabManager';
+
+const PREFIX = 'data-original-';
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:12]: maybe something like this, so its more clear what the prefix is?

```suggestion
const OG_PREFIX = 'data-original-';
```

File: src/utils/changeElementVOVisibility.js [hunk 5]
```diff
@@ -1,12 +1,44 @@
...
+const ARIA = 'aria-hidden';
+const TABINDEX = 'tabindex';
+
+function setOriginalValue(element, prop) {
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:16]: Maybe call it something like `preserveAttribute` or `cacheAttribute`?

`setOriginalValue` is very vague.

File: src/utils/changeElementVOVisibility.js [hunk 6]
```diff
@@ -1,12 +1,44 @@
...
+}
+
+function retrieveOriginalValue(element, prop) {
+  // get the cached property
+  const originalValue = element.getAttribute(PREFIX + prop);
+
+  // remove the prefixed attribute
+  element.removeAttribute(PREFIX + prop);
+
+  if (typeof originalValue === 'string') {
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:32]: I wonder why the check if its a `string` 🤔

File: src/utils/changeElementVOVisibility.js [hunk 7]
```diff
@@ -27,43 +59,46 @@ function iterateOverSiblings(el, callback) {
...
-      element.setAttribute(prop, originalValue);
-    } else {
-      // otherwise remove the attribute entirely.
-      element.removeAttribute(prop);
-    }
+  retrieveOriginalValue(element, ARIA);
+  retrieveOriginalValue(element, TABINDEX);
+
+  // make sure element's tabbable children are restored as well
+  const tabbables = element.querySelectorAll(`[${PREFIX + TABINDEX}]`);
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:96]: `TabManager.getTabbableElements` would not work here now, would it? 🤔 Because they are no longer tabbable.

**Review by dobromir-hristov — COMMENTED** [review]
I did a quick test and it looks to work good.

The code is OK, just left a few small comments.

File: src/utils/changeElementVOVisibility.js (hunk 7 above)
Comment (hqhhuang) [reply src/utils/changeElementVOVisibility.js:96]: Yup. If we use `getTabbableElements`, the only ones that would get restored would be the ones that are focusable by nature.

File: tests/unit/components/Asset.spec.js (hunk 2 above)
Comment (hqhhuang) [reply tests/unit/components/Asset.spec.js:188]: Yea I figured. These 2 are generated when I ran test.

Comment (hqhhuang) [reply tests/unit/components/Asset.spec.js:188]: Should we open a separate PR to fix these?

File: src/utils/changeElementVOVisibility.js (hunk 6 above)
Comment (hqhhuang) [reply src/utils/changeElementVOVisibility.js:32]: I tested this. If we call `element.getAttribute` on an attribute that doesn't exist, the return value is `null` and the type is therefore an object. 
In our case, the `cacheOriginalAttribute` ensures that even where there's no original value, we still set that attribute, just with an empty string. So in our case, the type of `originalValue` here should always be a string. However, if we had a bug where we somehow didn't set the attribute correctly, we would want to check the type here.

**Comment by hqhhuang** [comment]
@swift-ci test

File: src/utils/changeElementVOVisibility.js [hunk 8]
```diff
@@ -1,12 +1,44 @@
...
+import TabManager from 'docc-render/utils/TabManager';
+
+const OG_PREFIX = 'data-original-';
+const ARIA = 'aria-hidden';
+const TABINDEX = 'tabindex';
+
+function cacheOriginalAttribute(element, prop) {
+  // make sure that prop isn't cached already
+  let originalValue = element.getAttribute(OG_PREFIX + prop);
+  if (!element.hasAttribute(OG_PREFIX + prop) || originalValue === '') {
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:19]: Shouldnt we just check  if there is a cached attribute and only cache if there is not?

```suggestion
  if (!element.hasAttribute(OG_PREFIX + prop)) {
```

**Review by dobromir-hristov — COMMENTED** [review]
LGTM, left a small final question.

File: src/utils/changeElementVOVisibility.js (hunk 6 above)
Comment (marinaaisa) [reply src/utils/changeElementVOVisibility.js:32]: Thanks for the explanation, @hqhhuang , maybe then it's easier to do something like this?

````js

if (originalValue === null) return;

if (originalValue.length) {
  element.setAttribute(prop, originalValue);
} else {
  // otherwise remove the attribute entirely.
  element.removeAttribute(prop);
}
````

File: src/utils/changeElementVOVisibility.js [hunk 9]
```diff
@@ -1,12 +1,44 @@
 /**
  * This source file is part of the Swift.org open source project
  *
- * Copyright (c) 2021 Apple Inc. and the Swift project authors
+ * Copyright (c) 2022 Apple Inc. and the Swift project authors
```
Comment (marinaaisa) [inline src/utils/changeElementVOVisibility.js:4]: I believe we should do `2021-2022` instead of `2022` if we edit a file. :)

File: src/utils/changeElementVOVisibility.js [hunk 10]
```diff
@@ -1,12 +1,44 @@
...
+const TABINDEX = 'tabindex';
+
+function cacheOriginalAttribute(element, prop) {
+  // make sure that prop isn't cached already
+  let originalValue = element.getAttribute(OG_PREFIX + prop);
+  if (!element.hasAttribute(OG_PREFIX + prop) || originalValue === '') {
+    originalValue = element.getAttribute(prop) || '';
+    element.setAttribute(OG_PREFIX + prop, originalValue);
+  }
+}
```
Comment (dobromir-hristov) [inline src/utils/changeElementVOVisibility.js:23]: or maybe even:

```suggestion
function cacheOriginalAttribute(element, prop) {
  // make sure that prop isn't cached already
  if(element.hasAttribute(OG_PREFIX + prop)) return
  // cache the attribute
  const value = element.getAttribute(prop) || '';
  element.setAttribute(OG_PREFIX + prop, value);
}
```

File: src/utils/changeElementVOVisibility.js (hunk 6 above)
Comment (dobromir-hristov) [reply src/utils/changeElementVOVisibility.js:32]: Cant you just check for `if(originalValue)` ? Its always a `string` or `null`, so a simple `truthy` check should work. 

We could also probably store `OG_PREFIX + prop` as a var in the function, so we dont concat it 3 times.

File: src/utils/changeElementVOVisibility.js (hunk 8 above)
Comment (marinaaisa) [reply src/utils/changeElementVOVisibility.js:19]: Is there any reason why this function doesn't look like this?

```js
function cacheOriginalAttribute(element, prop) {
  // make sure that prop isn't cached already
  if (element.hasAttribute(OG_PREFIX + prop)) return;

  const originalValue = element.getAttribute(prop) || '';
  element.setAttribute(OG_PREFIX + prop, originalValue);
}
```
I don't understand why adding the originalValue as a let var before.

Comment (marinaaisa) [reply src/utils/changeElementVOVisibility.js:19]: so we wrote the same comment @dobromir-hristov 😆 https://github.com/apple/swift-docc-render/pull/427#discussion_r964889622
actually you wrote it first!

File: src/components/NavBase.vue [hunk 11]
```diff
@@ -317,16 +313,15 @@ export default {
     },
     onExpand() {
       this.$emit('open');
-      // lock focus
-      this.focusTrapInstance.start();
       // hide sibling elements from VO
       changeElementVOVisibility.hide(this.$refs.wrapper);
+      // focus on the toggle to prevent tabbing to links in the body
+      this.$refs.axToggle.focus();
```
[the remark this answers is not in the document]
Comment (dobromir-hristov) [reply src/components/NavBase.vue:319]: You click on another element, the axToggle is only when you focus navigate around :/

[the remark this answers is not in the document]
Comment (hqhhuang) [reply src/components/NavBase.vue:319]: This is needed to retain focus on the toggle when user expands Nav with mouse, then starts tabbing.

File: src/utils/changeElementVOVisibility.js (hunk 8 above)
Comment (hqhhuang) [reply src/utils/changeElementVOVisibility.js:19]: Because if the element has a cached attribute, but it's empty; we don't want to return here on this line:
`if (element.hasAttribute(OG_PREFIX + prop)) return;`

We would still want to set the attribute. Did I miss something?

File: src/utils/changeElementVOVisibility.js (hunk 10 above)
Comment (hqhhuang) [reply src/utils/changeElementVOVisibility.js:23]: Replied above under Marina's comment: https://github.com/apple/swift-docc-render/pull/427/files#r965056810

File: src/utils/changeElementVOVisibility.js (hunk 6 above)
Comment (hqhhuang) [reply src/utils/changeElementVOVisibility.js:32]: This is a good idea!
`if (!element.hasAttribute(OG_PREFIX + prop)) return;`

**Comment by hqhhuang** [comment]
@swift-ci test

**Comment by hqhhuang** [comment]
@swift-ci test

File: src/utils/changeElementVOVisibility.js (hunk 8 above)
Comment (marinaaisa) [reply src/utils/changeElementVOVisibility.js]: then what about `if (!!element.getAttribute(OG_PREFIX + prop)) return;` ?

**Comment by hqhhuang** [comment]
@swift-ci test

**Review by dobromir-hristov — APPROVED** [review]

File: src/components/NavBase.vue (hunk 11 above)
[the remark this answers is not in the document]
Comment (dobromir-hristov) [reply src/components/NavBase.vue:319]: I think we introduced a regression here....
